### PR TITLE
Add support for post-training dynamic quantization

### PR DESCRIPTION
**Pull request by danieldk**
This change adds the following:

- `TorchScriptShim`/`TorchScriptWrapper_v1` to wrap TorchScript.
- A spaCy CLI extension that provides the `quantize-transformer`subcommand.

Besides that there are some modifications to the models:

- Use a constant for the attention mask so that the encoder can be converted to TorchScript. For some reason, this can also not be a global constant, so we have to inline the constant.
- Annotate `TransformerEncoderOutput` with `torch.jit.script`, so that it works with TorchScript.
- Extend the model factories with an `torchscript` option to force construction of TorchScript wrappers.

### CODE REVIEW

File: curated_transformers/models/torchscript_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
+from typing import Any, Callable, Optional, Union
```
Comment (shadeMe) [inline curated_transformers/models/torchscript_wrapper.py:1]: Would this module be a good candidate to add/move to `thinc`?

File: curated_transformers/cli/quantize.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+from io import BytesIO
+from pathlib import Path
+import spacy
+from spacy import Language
+from spacy.cli import app
+from thinc.api import Model, PyTorchShim
+import torch
+from torch.nn import Embedding, Module
```
Comment (shadeMe) [inline curated_transformers/cli/quantize.py:8]:
```suggestion
from torch.nn import Module
```
Unused.

File: curated_transformers/cli/quantize.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+):
+    for pipe_name, pipe in nlp.components:
+        # We only quantize curated transformers, other models may not
+        # be able to deal with quantization and/or TorchScript.
+        if not isinstance(pipe, Transformer):
+            continue
+
+        model = pipe.model.get_ref("transformer")
+
+        if model.name != "pytorch":  # Probably already quantized.
```
Comment (shadeMe) [inline curated_transformers/cli/quantize.py:52]: We could be explicit here and compare it to `pytorch_script` instead.

File: curated_transformers/models/torchscript_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        self.cfg = msg["config"]
+        filelike = BytesIO(msg["model"])
+        filelike.seek(0)
+        self._model = torch.jit.load(filelike, map_location=device)
+        self._model.to(device)
+        self._grad_scaler.to_(device)
+        return self
+
+
+def TorchScriptWrapper_v1(
```
Comment (shadeMe) [inline curated_transformers/models/torchscript_wrapper.py:68]: Would it make sense to register this entry-point in `setup.cfg`?

File: curated_transformers/models/torchscript_wrapper.py (hunk 1 above)
Comment (danieldk) [reply curated_transformers/models/torchscript_wrapper.py:1]: Yeah, this should eventually go into Thinc, once we are happy with the class.

Comment (shadeMe) [reply curated_transformers/models/torchscript_wrapper.py:1]: I'll open an issue to keep track of it.

File: project/project.yml [hunk 5]
```diff
@@ -61,18 +62,31 @@ commands:
...
       - "python -m spacy evaluate --gpu-id ${vars.gpu} training/${vars.treebank}/model-best corpus/${vars.treebank}/test.spacy"
+      - "python -m spacy evaluate training/${vars.treebank}/model-best-quantized corpus/${vars.treebank}/test.spacy"
     deps:
       - "training/${vars.treebank}/model-best"
+      - "training/${vars.treebank}/model-best-quantized"
       - "corpus/${vars.treebank}/test.spacy"
 
   - name: "evaluate-dev"
-    help: "Evaluate the parser model on the test corpus."
+    help: "Evaluate the pipelinel on the dev corpus."
```
[the remark this answers is not in the document]
Comment (danieldk) [reply project/project.yml]: Fixed.

File: curated_transformers/cli/quantize.py (hunk 2 above)
Comment (danieldk) [reply curated_transformers/cli/quantize.py]: Fixed, but slightly differently. I also import `Linear` now and replaced the fully qualified `torch.nn.{Embedding,Linear}` later in the file.

File: curated_transformers/cli/quantize.py (hunk 3 above)
Comment (danieldk) [reply curated_transformers/cli/quantize.py]: 👍 much better

**Review by shadeMe — APPROVED** [review]
